lib/http/2/stream.rb

```diff
@@ -97,14 +98,18 @@ def receive(frame)
 
       case frame[:type]
       when :data
-        window_size = frame[:payload].bytesize
-        window_size += frame[:padding] || 0
-        @local_window -= window_size
+        delta_size = frame[:payload].bytesize
+        delta_size += frame[:padding] || 0
+        @local_window -= delta_size
         emit(:data, frame[:payload]) unless frame[:ignore]
 
+        adjust_recv_window_size(delta_size)
         # Automatically send WINDOW_UPDATE,
         # assuming that emit(:data) can now receive next data
-        window_update(window_size) if window_size > 0
+        if @recv_window_size > 0 && @recv_window_size >= (@local_window / 2)
```

Comment:
We should document what's happening here, alongside the tradeoffs... E.g we're reducing number of window_update frames, but this does create cases where the server might be blocked because it exhausts the negotiated window and is waiting for us to bump it.